### ISSUE TITLE
Recently introduced bug in PACKFILENAME

### ISSUE THREAD

**Issue by fghalasz** [Interlisp/medley#1685]
Here is a PACKFILENAME call from most recent Master branch:
![master_now](https://github.com/Interlisp/medley/assets/1251539/82769726-c8c8-4c27-8b9e-383eee013146)
Note the empty top level directory

Here is the same PACKFILENAME call using the Master branch as of February this year (specifically, 2024-02-28)
![master_240228](https://github.com/Interlisp/medley/assets/1251539/2ebdd3e1-d956-4c24-b5b5-8e20c943bc5d)
The top level directory is correct here - kind of.  But if you then add a filename or subdirectory, the loss of the closing angle bracket gets papered over.

Assigning to Ron since I think he recently made some changes to PACKFILENAME.

**Comment by nbriggs** [comment]
The current behavior is correct - the /medley is a *file* named medley not a *directory* named medley.

**Comment by nbriggs** [comment]
```
2/13_ (PACKFILENAME 'HOST 'DSK 'DIRECTORY 'MEDLEY )
{DSK}<MEDLEY>
2/14_ (PACKFILENAME 'HOST 'DSK 'DIRECTORY 'MEDLEY 'BODY 'FOO.TXT )
{DSK}<MEDLEY>FOO.TXT
```

**Comment by fghalasz** [comment]
Actually /medley in the Unix world could be either a file or a directory - can't tell from the pathname, as is the case for any Unix pathname that doesn't end in /.   This leads to an ambiguity when translating into {DSK}<> path notation since /medley could be translated equally correctly into either `{DSK}<>medley` or `{DSK}<medley>`.   To complicate things even more, until recently in /medley was translated into `{DSK}<medley` (without the closing bangle bracket).  Is this is even a well-formed pathname?

Note that most of the system seems to operate using the '{DSK}<medley` (no closing angle bracket) notation.
See the following examples:
![case1](https://github.com/Interlisp/medley/assets/1251539/82464624-d426-459e-a24b-076d8621ff4f)
 
![case2](https://github.com/Interlisp/medley/assets/1251539/607891ef-ab29-4017-bf0d-7ba99f7757b1)

![case3](https://github.com/Interlisp/medley/assets/1251539/410e7b0e-40fb-4e85-b820-2755269b9e71)

Also see the following:
![case4](https://github.com/Interlisp/medley/assets/1251539/90b2ad2e-8f74-4431-b56e-29d2b39dd75b)


My take away from all of this is that the "correct" translation of \medley is what my PACKFILENAME examples were doing BEFORE the recent changes:  /medley translates to `{DSK}<medley` (no closing angle bracket).    The change to this in the current master is going to lead to lots of weird bugs like the Cygwin failures Steve K discovered - if only because we've been operating the no closing angle bracket way for some time and have it baked into lots of places in the code.

**Comment by fghalasz** [comment]
Also note that the problematic change is in UNPACKFILENAME.STRING and not PACKFILENAME.
In the master from 2024-01-15:
![case2 1](https://github.com/Interlisp/medley/assets/1251539/42fb7d93-8081-480a-90d6-259d3497debf)

In the current master:
![case2 2](https://github.com/Interlisp/medley/assets/1251539/ecf165d4-c7a3-488c-9563-401eb97cbba0)

**Comment by nbriggs** [comment]
However no closing angle bracket is a syntax error in the scheme where you're using `<...>` rather than `/`.

**Comment by rmkaplan** [comment]
The change was in UNPACKFILENAME, not PACKFILENAME, to deal with a NIL directory that was causing a crash.

(UNPACKFILENAME ‘/medley) now returns (DIRECTORY NIL NAME medley), i.e. treats it as a file.

For (UNPACKFILENAME ‘/medley.xxx) it returns (DIRECTORY NIL NAME medley EXTENSION xxx)

Would you have wanted (DIRECTORY medley.xxx) in that case?

**Comment by nbriggs** [comment]
The change in `UNPACKFILENAME.STRING` is correct - it was an error to return a directory with a syntax error: `"<"` rather than an empty directory, `""`

If you meant "medley" (or "/medley") to be a directory then say that to PACKFILENAME - I think you'll get the right thing

![image](https://github.com/Interlisp/medley/assets/8138841/c367d374-fb4d-42ac-a9b2-a8a672e162e6)

**Comment by fghalasz** [comment]
Here's another problematic entailment of the new UNPACKFILENAME

(PACKFILENAME (UNPACKFILENAME 'XXX)) should return XXX, right?

But, in the current master:
```
17_ (PACKFILENAME (UNPACKFILENAME '{DSK}<>foo))
{DSK}foo
```
which is not correct.

But in the 2024-01-15 master:
```
4_ (PACKFILENAME (UNPACKFILENAME '{DSK}<>foo))
{DSK}<>foo
```
which is correct.

The new UNPACKFILENAME is broken because the whole Interlisp pathname scheme really didn't take into account files directly under the root - it was designed with the assumption that the top-level would always be a directory - which in most cases it is - expect when it isn't as in the Cygwin issue.

**Comment by fghalasz** [comment]
Back to the /medley issue:  there is no CORRECT translation of /medley to {DSK}... because the Unix and Interlisp pathnames are not isomorphic - especially when it comes to the top-level directory.  So we have to choose on a pragmatic basis - what works.

Someone clearly thought this thru at some point and came up with the "no closing angle bracket" scheme.  It may not be a syntactically correct Interlisp path name - but it was built into lots of parts of the system as I have shown in the examples above.

In particular, you can use a "no closing angle bracket" pathname in both OPENSTREAM and CNDIR - meaning the path will be interpreted as a file or a directory depending on the context (as is the case in Unix).

And until the recent change, it was also built into UNPACKFILENAME (and PACKFILENAME).

**Comment by fghalasz** [comment]
@rmkaplan
(UNPACKPATHNAME '/medley) should return (DIRECTORY < NAME medley)
(UNPACKPATHNAME '/medley.xxx) should return (DIRECTORY < NAME medley EXTENSION xxx)
which is what it did before you made the change to UNPACKFILENAME.

Also (UNPACKFILENAME '<>medley) should return (DIRECTORY > NAME medley).

Returning NIL as the value of the DIRECTORY in any of these cases breaks a bunch of other stuff in the system.

**Comment by rmkaplan** [comment]
So you want < to be a directory all by itself, if a > doesn’t appear later in the string?

Usually the open and closing brackets of a directory are stripped off, so that <abc>xyz>qrs>  shows up as abc>xyz>qrs.  But in the odd case here, when there is no later >, you want the < to sit in the directory field all by itself?

I think there was a discussion earlier about the fact that UNPACKFILENAME doesn’t currently distinguish empty from missing in its return (NIL in both cases), whereas UNPACKFILENAME.STRING at least can distinguish between missing (NIL) and empty “”.  It probably wouldn’t help this issue, but perhaps UNPACKFILENAME should return the empty atom instead of NIL

**Comment by fghalasz** [comment]
The DIRECTORY should be "\<" if and only if the Unix pathname is of the form `/xxxx` (with no additional slashes) or the Interlisp pathname is of the form` {DSK}<xxx` (with no additional angle brackets).

Also DIRECTORY should be ">" if the Interlisp pathname is of the form `{DSK}<>xxx`, with no additional angle brackets.

This is exactly the way that UNPACKFILENAME worked before the recent changes.

There is clearly "special" code in the system to handle Interlisp paths of the form `{DSK}<xxxx`.  For example, OPENSTREAM will treat this as a file reference (to the file /xxxx) equivalent to `{DSK}<>xxxx`   But CNDIR and DIRECTORYNAME will handle this as a directory reference (to the directory /xxxx) equivalent to `{DSK}<xxxx>`.  (I have yet to track down said "special" code - but the behavior of these functions is clear.)

Pretty clear that this "special" code is there to handle the difference in how Unix and Interlisp handle any top-level files/directories that are directly under the filesystem root.

**Comment by rmkaplan** [comment]
In the Venue sysout, UNPACKFILENAME gives the same result for  {DSK}/medley  {UNIX}/medley  {DSK}<medley {UNIX}<medley
      HOST xxx DIRECTORY <  NAME medley

I.e., UNPACKFILENAME is a Lisp function, it’s behavior is not provided by a function in the file device.

{UNIX}>medley and {DSK}>medley both give DIRECTORY > NAME medley.

But at least the function DIR shows the same files for UNIX and DSK
   DIR <
   DIR >
   DIR /
and in all cases the top level directory is printed as either {UNIX} < or {DSK}<.

In other words, at least for DIR, the behavior is the same no matter how the top-level is specified.

Maybe the DIRECTORY field should be returned as < even for the 2 > cases, as the canonical form of the top-level directory no matter how specified.

**Comment by nbriggs** [comment]
I'm not sure that there's evidence of it having been thought through.  I think it was hacked together so that it sort of worked as long as you didn't try to push the boundaries.  I'm confused as to why MEDLEYDIRS code and/or startup scripts would be getting anything other than "proper" directory names?  If it's working with something it knows is a directory it probably shouldn't be passing something as a `BODY` and hoping that the final component is interpreted as something other than a filename.

Re: DIRECTORYNAME - that's not just parsing, it's checking the existence of the directory, so your examples above depend on exactly what your filesystem looks like and where you're connected. Which means that:

```
PWD
"{DSK}<Users>"
(DIRECTORYNAME "Briggs")          relative to the default path, exists
"{DSK}<Users>Briggs"
(DIRECTORYNAME "nobody")        relative to the default path, doesn't exist
NIL
(DIRECTORYNAME ">Briggs")       relative to the default path,  {dsk}<Users>>Briggs OK as multiple > collapse
"{DSK}<Users>Briggs"
(DIRECTORYNAME ">>>Briggs") as above
"{DSK}<Users>Briggs"
(DIRECTORYNAME "<Briggs")       < pulls up to top, {DSK}<Briggs> does not exist
NIL
(DIRECTORYNAME "/Users/Briggs") absolute path on {DSK}, leading / is like leading <
"{DSK}<Users>Briggs>"
(DIRECTORYNAME "/Briggs") absolute path, {DSK}<Briggs> does not exist
NIL
(DIRECTORYNAME "/Users////Briggs")  absolute path, multiple / collapse
"{DSK}<Users>Briggs>"
```

**Comment by fghalasz** [comment]
@nbriggs  You are right that "thought out" is way too strong an expression.  More like futzed with various pieces until they basically got it working OK all around.

My point is that unless we want to truly overhaul the whole pathname handling throughout the system, lets just leave all of the various pieces in place unless there is a clear need to change them.  The reason being that small changes to one piece is likely to have unpredictable consequences on other pieces.

The changes to how UNPACKFILENAME behaves with respect to these top-level files/directories is a good example.   If Ron really needs to change the behavior of UNPACKFILENAME *for these cases* to fix his bugs, then we'll just have to deal with the consequences.

But if the changes *for these cases* is just an accidental result of fixing a bug with some other cases - then let's leave the behavior with respect to these cases alone so we don't have to spend time re-futzing to deal with the side effects of unnecessary changes.

**Comment by masinter** [comment]
I think in order for us to run Common Lisp programs developed elsewhere, we need to fix the handling of Common Lisp pathnames to be compatible, not particularly with CLtL2 or ANSI Common Lisp but the requirements of compatibility with popular Common Lisp implementations, at least sufficient to be able to run 
   [ASDF Manual (common-lisp.dev)](https://asdf.common-lisp.dev/asdf.html) and [Quicklisp](https://www.quicklisp.org/beta/) .  Note that the Medley implementation of pathnames stores the file 'name' as a parsed property list and parsing happens when the pathname is created instead of being manipulated.
we can then follow a popular CL and not debate "the right thing".

**Comment by rmkaplan** [comment]
I'm still trying to make sense of this, here's another try.

 The current UNPACKFILENAME produces different results for
  /foo  → DIRECTORY NIL NAME foo        (NIL indicating the top-level directory, same result with <, >, or <> for /)
  foo   → NAME foo                                    (no top-level directory)
but PACKFILENAME loses the distinction by converting both of these to just foo.  

The problem is that when UNPACKFILENAME recognizes any of the several ways of specifying the top-level directory, its result doesn't represent the difference between no directory and top-level directory in a way that PACKFILENAME can properly interpret. 

Now, PACKFILENAME produces the same result (<>foo) for any of
     DIRECTORY < NAME foo
     DIRECTORY > NAME foo
     DIRECTORY <> NAME foo
     DIRECTORY // NAME FOO
and repetitions of various sorts.

So, would it be sufficient to pick just one of these representations as the canonical way for UNPACKFILENAME to indicate that the top-level directory was specified?  I.e., why try to make a distinction that PACKFILENAME doesn't care about?

Given that PACKFILENAME currently canonicalizes all of these to <>foo (and nobody wants to mess with PACKFILENAME), I suggest changing UNPACKFILENAME so that it always produces  DIRECTORY <> however it recognizes a top-level directory.  On this proposal, all of the different <,>,/ formats would map to DIRECTORY <> NAME foo.

**Comment by rmkaplan** [comment]
However...
What should happen if <> appears in the middle of a directory sequence:

    <abc<>>xyz>foo  →  DIRECTORY abc>xyz NAME foo  or  DIRECTORY abc<>>xyz  NAME foo

   It would be the first, on the analogy with /abc//xyz/foo.  Currently it produces the second.

**Comment by nbriggs** [comment]
I loaded the new ADIR.LCOM, and I get, from `(UNPACKFILENAME "{DSK}<abc<>>xyz>foo")`
```
(HOST DSK DIRECTORY abc<>xyz NAME foo)
```
and the actual file name in the Unix file system would be `/abc</xyz/foo` with the `<` literally in the directory name (were it not a protection violation).

`(APPLY 'PACKFILENAME (UNPACKFILENAME "{DSK}<abc<>>xyz>foo")` is `{DSK}<abc<>xyz>foo`, since the extra `>` is collapsed out.

**Comment by rmkaplan** [comment]
This is the behavior of internal directory brackets in the Venue sysout, and I think I just tried to replicate that when I did the rewrite.  That’s not a change in the ADIR in the new PR.

An alternative principle would be to treat all directory-internal brackets, those between the initial < or / and the final directory-ending  > or /, as if they were slashes, and then do what a Unix shell would do:  eliminate any duplicates.

So {DSK}<abc<>>xyz>foo     would be treated as
      {DSK}/abc///xyz/foo          and result in
      DIRECTORY abc>xyz NAME foo

**Comment by nbriggs** [comment]
Another thing I note is that when communicating file names from Lisp to Maiko, the files are normalized to contain ONLY `>` as path component separators, with duplicate `/` and `>` removed before they're passed to Maiko, and the Maiko code is responsible for conversion of the "Lisp" filename to the "Unix" filename.  Maiko's conversion is always to an absolute path name (never relative to the process's Unix working directory) since it is the Lisp code that maintains and applies the `*DEFAULT-PATHNAME-DEFAULTS* and `\CONNECTED.DIRECTORY`  to filenames before passing them to Maiko.  It surprised me to discover that the Maiko code to call `chdir()` from the DSK/UNIX device code is never called.

**Comment by rmkaplan** [comment]
@nbrigg notes that Maiko only receives a filename that has already been canonicalized at the Lisp level, presumably by different combinations of UNPACKFILENAME, PACKFILENAME, ADD.CONNECTED.DIR, etc.  I'll further note that the Lisp-only CORE device also traffics in names that it expects to have been canonicalized.

I want to call attention again to the straw-man roposal I made in an earlier comment:  A simple way of specifying and understanding what UNPACKFILENAME produces for oddball combinations of directory brackets < > /, is that < and > are first mapped to / in every position, and those / are then treated the way Unix would treat / in that position: 
  - duplicate / would be removed
  - top-level directory would be recognized as an initial / not followed by any other /
Then the Medley output would be constructed by
  - representing with <  the top-level (/) directory 
  - otherwise stripping an initial /, and always stripping a final /
  - and representing by > the / that remains to separate each internal subdirectory
  
So  
     <>/>abc//>xyz/><<qrs  would show up as DIRECTORY abc>xyz NAME qrs
     <>/<<abc.foo would show up as DIRECTORY < NAME abc EXTENSION foo
This would mean that none of these directory brackets, not even <, would show up in the name field.  In contrast, if < isn't included as a downstream slash-equivalent (the easiest implementation), the first example would result in DIRECTORY abc NAME <<qrs.

**Comment by masinter** [comment]
What about pseudo-hosts and host names in general.
is {MEDLEY}sources/ADIR.LCOM => {MEDLEY}<sources>ADIR.LCOM ?

**Comment by rmkaplan** [comment]
This all happens after the end of the host has been identified, if there is one.  Pseudohost doesn't interact at the level of file-name manipulation, a pseudohost is just another host (like a different IFS server).

**Comment by fghalasz** [comment]
@rmkaplan  Not sure I understand the conversion back to Medley rules.
By
  - representing with < the top-level (/) directory
  - otherwise stripping an initial /, and always stripping a final /

Why strip off any initial / - you need some way of differentiating absolute from relative pathnames.  The way I read this, /foo/bar/baz and foo/bar/baz would both result in DIRECTORY foo>bar> NAME baz.

Or do you mean?
  - first replace any leading / with <
  - then strip off any final /

In which case, /foo/bar/baz would result in DIRECTORY <foo>bar>, while foo/bar/baz would result in DIRECTORY foo>bar>, which is what I think we want.

**Comment by fghalasz** [comment]
Also FWIW, "NAME <<qrs", i.e., a file or directory named `<<qrs`,  is allowed (although not very smart) in both 'standard' Linux (ext) filesystems and then 'standard' MacOS (HFS+) filesystem.  It is not allowed on any Windows-y filesystem (NTFS, exFAT, etc) even if mounted on Linux or MacOS.

Under your scheme, Medley could not read/create any such files?

Seems OK to me, but we should document such restrictions.  Or maybe just claim to only handle files that follow POSIX filename conventions.

**Comment by rmkaplan** [comment]
UNPACKFILENAME has always made that distinction in another way (FWIW?):

(UNPACKFILENAME “foo/bar/baz”)  produces SUBDIRECTORY foo/bar NAME baz
(UNPACKFILENAME “/foo/bar/baz”) produces         DIRECTORY foo/bar NAME baz

**Comment by nbriggs** [comment]
But (UNPACKFILENAME "{DSK}foo/bar/baz") produces (HOST DSK RELATIVEDIRECTORY foo>bar NAME baz) while if you leave off the {DSK} it does as you note.
